###This file will walk you through the different population genomics steps and requirements of the analysis for the results published in Renaut S, Rowe HC, Ungerer MC, Rieseberg LH. 2014. Genomics of homoploid hybrid speciation: diversity and transcriptional activity of long terminal repeat retrotransposons in hybrid sunflowers.  Philosophical Transactions of the Royal Society B: Biological Sciences.  http://dx.doi.org/10.1098/rstb.2013.0345

###You may also look at the different scripts to see more details.
###Use at your own risks...
###It doesn't do the expression analysis of the data described in the manuscript.

### Sebastien Renaut - February 2013 ###
### sebastien.renaut@gmail.com ###

###
###required directories
###
#You need a main directory were are your files are and analysis will be done (here this is /media/seb_1TB/transposon)
#In this directory you'll have different subdirectories
#data/ : contains the data as _1.fq/_2.fq files or sub directories which contain several fastq.gz (eg. : "Goblinvalley_AGTCAA_L006_R1_001.fastq.gz", where 001 indicates it is the first subfile, and R1 means it is pair 1 or 2.) files which need to be unzipped and concatenated.
#results/ : will contain the main result table
#alignments/: will contain the alignments generated by bwa.
#Rcode/ : contains the Rcode required for analysis 
#reference/ : contains the reference fasta file (here HA412_trinity_noAltSplice_400bpmin.fa) and a list of the .fq sequences ("hybrid_species") in the proper format ("path\to\sequence"	"species"	"ill"	"sanger").
#mpileup/ : will contain temporatory files generated by mpileup. 
#mpileup/cmd : will contain temporatory files generated by mpileup.

###
###required files & programs
###
#Unix based machine (Mac or Linux), >2-core machine.
#reference transcriptome (.fasta file and split file), sequence files (.fastq), list of 
#R (â‰¥ 2.15.0, http://www.r-project.org/)
#bwa (http://bio-bwa.sourceforge.net/)
#samtools (http://samtools.sourceforge.net/)
#working copy of java (http://www.java.com/en/download/index.jsp)
#trimmomatic (http://www.usadellab.org/cms/index.php?page=trimmomatic)
#vcfutils.pl (from samtools, but copy it to your Rcode directory)

###
### Running analysis
###
#analyses were all ran from the command line (although one could run them directly from the R console too)

#trims files with trimmomatic
nohup ./Rcode/trimmomatic_cluster.R 1 18 >trim.log& #here numbers are for the files you want to process.

#align with bwa and sort with samtools
nohup ./Rcode/alignments_cluster_v2.R 1 18 >align.log&

#get some alignment statistics 
nohup ./Rcode/alignments_stats.R 1 18 >stats.log&

#run mpileup to get SNPs (there is a bug in this script which I only noticed later. At lines 102 and 105, "A" was be replaced by "L")
nohup ./Rcode/snp.R all 8 snp >snp.log&
nohup ./Rcode/snp.R all 8 snp_stats >snp_stats.log&
 
#parse and trim the SNP file, get some stats (including Fst)
nohup ./Rcode/snp_II.R deserticola 8 snp >snp_des.log& #here number is for CPUs
nohup ./Rcode/snp_II.R anomalus 8 snp >snp_ano.log& 
nohup ./Rcode/snp_II.R paradoxus 8 snp >snp_par.log& 
nohup ./Rcode/snpII.R deserticola 16 snp_stats >snp_statsII.log&

#calculate nucleotide diversity (pi)
nohup ./pi.R pi.log&

#calculate proportion of non-synonymous sites fized by selection (alpha)
nohup ./ORF_MKtest.R 1 7 >ORF_MKtest.log&

#call a few (1000 genes) SNP in order to get an input file for SPLITSTREE.
nohup ./NJtree.R all 6 >NJtree.log&

###You are done! You may need to remove a few unwanted log files, unpaired trimmomatic files, .bam.bai files and do some graphs. Also it is likely that the code will break because you do not have the proper directories.

###Any questions?

###Feb 2014 updates
#Of course, as the manuscript progressed, the code become more bloated and new analyses had to be done...
#I tried to keep the script organised, but it got a little out of hand
#In addition, many of the expression results were perfomed by Heather Rowe and are not documented here...










